text/0000-let.md

```diff
@@ -0,0 +1,265 @@
+- Start Date: 2017-01-14
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Introduce a `let` keyword to bind values to names within a handlebars
+template.
+
+Block form:
+
+```handlebars
+{{#let "Milk" "Cereal" as |one two|}}
+  Breakfast:
+  1. {{one}}
+  2. {{two}}
+{{/let}}
+```
+
+or inline:
+
+```handlebars
+{{let one="Milk" two="Cereal"}}
+
+Breakfast:
+1. {{one}}
+2. {{two}}
+```
+
+# Motivation
+
+The primary method for supplying named values to a template is via
+properties on its backing scope (either a controller or a
+component). These properties are specified in JavaScript which, while
+as a mechanism is very well understood by the community, also comes
+with several drawbacks.
+
+The first is that it spreads the requisite knowledge to understanding
+a template across two places. In other words, it is impossible to understand
+what exactly it does without simultaneously holding both the `.hbs`
+and `.js` files together and continuously merging them with your
+mind. The lack of this overhead is one of the commonly cited strengths
+of JSX over Handlebars.
+
+Another is that computed properties are error-prone because of their
+requirement to explicitly enumerate dependencies. If the computation
+of a property has incorrectly specified dependent keys it will break
+in mysterious ways that usually involve a sudden and hard-to-reproduce
+lack of reactivity to changing inputs.
+
+In order to avoid these pitfalls, developers have begun moving more
+and more computation into templates, and the rise of popular of addons
+like `ember-truth-helpers`, `ember-math-helpers` and
+`ember-composable-helpers` are strong evidence of this pressure.
+
+For example:
+
+```hbs
+<button disabled={{or isPending isError}}>Submit</button>
+```
+
+This solves the problem of proximity by placing the derivation of a
+value right next to where it's used. There's no question about how the
+disabled attribute is computed. It also solves the dependency problem
+because the dependencies of an expression are implicity enumerated
+merely through the act of expressing it.
```

Comment:
I see this is as a problem generally with using shared mutable state. Unfortunately, this is still the common Ember practice, and in practice what you see all the time is users only specifying the base object and not attributes in their computed properties.

The way we write apps today with an immutable style, the `fullName` helper you describe works just fine, because a change to `user.firstName` means that the reference to `user` changes as well.